### ISSUE TITLE
Ops suggestion: change table schema

### ISSUE THREAD

**Issue by rix0rrr** [Felienne/hedy#243]
Looking at the Dynamo config, I have a couple of recommendations.

## Programs table schema

Right now the table schema looks like this:

```
programs:
KEY: id -> { ...all program data }
INDEX: username -> { ...all program data }
INDEX: name -> { ...all program data }
```

Couple of notes:

* You don't HAVE to project all the attributes into every index (keep in mind DynamoDB indexes are clustered indexes). Right now, you are basically duplicating all the data from the `programs` table 3 times, once for every index. However, you can also project a subset of attributes, and this will save storage over time. For example, you probably have the username index there to look up all the programs for a certain user. You don't HAVE to have the program code in that index (which you do now). The index could equally well have been:

```
INDEX: username -> { name, id }
```

That gives you all the information you need for your query, and saves space (the program text is for example not duplicated, nor timestamps and other metadata you probably don't care about).

* What's the point of the `name` index? Since names don't need to be unique, I can't imagine you ever look up a program by `name`, so why have an index on this at all? Is that not just a wasted index? (= wasted space, = wasted capacity, = wasted $$$).

* Don't forget about sort keys! I imagine if you laid out the key schema and indexes slightly differently, you could still do all you needed to without much trouble. 

```
# Proposed new schema
programs
KEY: user (partition) -> id (sort) -> {...all program data }
INDEX: id -> { user }
```

If you use this schema, aggregate operations become a lot simpler:

* *get_programs_for_user* becomes a single query operation that doesn't need an index anymore (saves space, saves $$$).
* *count_programs_for_all_users* either becomes a single full table scan, or a scan of the `id -> { user }` index if you want to read less data.

There is a downside:

* *get_program_by_id* now needs to first look up `user` in the index, and then look up the actual program using the key pair of `(user, id)`. So looking up a single program given just its id becomes slightly more expensive now (2 lookup operations instead of one, but both of them are <5ms so total time will still be <~10ms). If that's a concern, you could solve for that by always including the username into the ID, and if you're concerned about privacy leaks you could resort to a user ID instead of a user name.

Still think this slightly adjusted schema will be worth the savings in space.

## Users table schema

Similarly for the `users` table:

```
users:
KEY: username -> {...user data}
INDEX: email -> { ...user data}
```

Duplicates all user data across the table and index. The following:

```
users:
KEY: username -> {...user data}
INDEX: email -> { username }
```

Would allow you to do everything you realistically need to, but saves space.

I guess, not a big deal at current data volumes, but let's not design for failure.

**Comment by fpereiro** [comment]
HI @rix0rrr ! Thanks for all the suggestions! My knowledge of DynamoDB is very basic, so I really appreciate the feedback. I will work on implementing your recommendations and keep you updated.

**Comment by Felienne** [comment]
@fpereiro Just checking, is this still ongoing work or planned work or can it be closed?

**Comment by fpereiro** [comment]
Hi @Felienne ! Indeed this is still pending and on my list :)

**Comment by fpereiro** [comment]
Hi @rix0rrr ! I finally took the time to review your suggestions and the existing indexing.

I wasn't aware that projecting all attributes means copying all the data on the index - I assumed that the index simply worked as a pointer to the one copy of the row/entry. Good to know!

You're absolutely right about the `name` index being useless. I think at some point we thought about scanning programs by name but that never materalized and indeed it doesn't seem to be useful. I will remove the index.

Regarding the other suggestions, if I understood them correctly, they amount to having each index (`email` for the `users` table and `username` for the `programs` table) to project only one field (`username` for the `users` table and `id` for the `programs` table). The upside for this would be to reduce by half the amount of space we use in DynamoDB (since the data is duplicated by each index and we have one index on each table, it'd be a 2x). But the downsides (some of them you mention) would be the following:
- More code: when searching with an index, more code in the utility functions to retrieve the full object.
- More requests: when loading all the programs for a given user (something we now do with every request to the code section), we'd have to perform N requests to load the actual programs; and when loading all the users in the admin, the same would happen.

Now, requests are very very cheap (read requests are 0.3 USD per million), but so is storage (0.3 USD per GB). The main drawback here would be that all these extra requests would slow down our application - Python works synchronously by default, so all these 5 or 10ms latencies would add up, especially on our current single core setup. So I think that any cost savings we perform by reducing the amount of data stored on DynamoDB would be more than offset by the need to have more CPU cores to perform these requests - and perhaps even by the amount of read requests. For reference, we currently have 3.2MB in our `programs` table and 335KB in our `users` table.

I don't think I understood well your suggestion to use a sort key to get the programs from a certain user, so perhaps I'm missing something big there.

Let me know what you think, especially regarding the cost of adding (synchronous) requests to Python.

**Comment by rix0rrr** [comment]
You are right that we need to think about latencies, and you are also right that data size is not really an issue right now. Perhaps it's indeed best to leave the tables as-is. Since we switched to on-demand pricing (instead of provisioned capacity) costs have gone to virtually nil anyway (<$1.00/mo), so there's no real point optimizing.

If anything, we should be thinking about denormalizing to trade some GB and $$$ for speed.

So I fully support not doing anything about this after all.

----

Some point-by-points:

> Downside [...] More code: when searching with an index, more code in the utility functions to retrieve the full object.

To me, that is not an argument one way or the other. Code by itself doesn't cost much[0], any objectives we'd want to achieve with that code are worth more.

> More requests: when loading all the programs for a given user (something we now do with every request to the code section), we'd have to perform N requests to load the actual programs

Wait what? Why would we load all actual programs? Only thing we need to load is the LIST of programs to show the user, and we don't need the actual program code. We only need the list of names and identifiers to properly render a page, don't we?

If we are also loading the actual program code, I would say that's actually wasteful: we're loading and transferring bytes (costing time!) that we're not using.

> I don't think I understood well your suggestion to use a sort key to get the programs from a certain user, 

DynamoDB acts as both a HASH TABLE and a SORTED ARRAY at the same time, so you can get the benefits of both. Specifically, 
it acts as a hash table pointing to sorted arrays (if you specify a "sort key"). if you pick the keys right, you can remove the need for a separate index altogether.

It works like this:

* The hash table part of DynamoDB is indexed by the "hash key" (sometimes called "partition key"). You're already familiar with this.
* The second level depends on whether or not you have a "sort key". 
  * If you DON'T have a sort key, every "bucket" in the hash table is a single record.
  * If you DO have a sort key, every bucket in the hash table is conceptually an array of records, which are ordered by whatever you specify as the "sort key". 

This has two advantages:

1. You can conceptually index into an array given a hash key (same as an index gives you, but without the need for an additional index)
2. If the "ordering" actually has meaning, you can do queries like "give me all records AFTER this key, BEFORE this key", etc.

This means that if we use the following as key schema:

* Hash key: userid
* Sort key: programid (or program name, or program timestamp (or maybe a combination of `timestamp + program id` to ensure uniqueness)

Then just a single query into the `program` table by `userid` is enough to retrieve all programs for a single user.

-------

> [0] I mean, yes I know it does. But not at the level of work we're talking about here.

**Comment by fpereiro** [comment]
Hi @rix0rrr !

Thank you for the detailed explanation about DynamoDB sort keys. I finally get it :). It's very interesting to have a data structure that works both as a hash and an array.

Regarding the retrieval of all programs:
- In an upcoming PR (https://github.com/Felienne/hedy/pull/372) we automatically load your saved programs for a combination of an adventure+level. So, if you are in level 1, and you go to Dice, and you already had a program saved for Dice+Level 1, said program will be automatically loaded. We also do this for programs saved for "plain" levels. In this case, we need the entire program, not just some fields, because we can potentially. A more efficient way to do it would be to retrieve, for the requesting user, the latest program for a given combination of adventure + level - this would require two extra indexes, one for `adventure_name` and one for `level`. 
- In the same PR, if the user lands on /hedy/adventures/ADVENTURE_NAME, we retrieve the saved programs to determine the saved program for that adventure with the highest level, so that users can resume where they left off. The query could be improved in the same way than the previous one just mentioned. Here there's indeed inefficiency in bringing the program body, but it is important to be mindful that the average program is 20-40 bytes.
- The programs view has to retrieve all the programs, since all of them (including text) are shown to users. Here there's no possible improvement.
- When saving programs, we also retrieve all the user's programs, to see if they have a program with the same name. If they do, we add a `(1)` to the name - and if there's a program with that name and a `(1)`, we increment it to `(2)` in the saved game of a new program. To avoid this, we would have to have an index for program names (that was the original rationale for doing this, until it became clear it was probably quite clunky) and incrementally query for the `(1)`, `(2)`, etc.
- In general, the average amount of programs per user is 6, and the average size of a program is 400 bytes (as stored in DynamoDB). Unless both numbers grow by at least an order of magnitude, I think that any improvements will be marginal in their cost/performance impact. This may well change in the future, so it's good to keep an eye on it.

**Comment by rix0rrr** [comment]
I would feel more comfortable if we had a clear query log, to validate all the assumptions and finding the hot spots based on data.

Right now we still have queries taking a couple of seconds (!) to service, and that is just not good enough. Our goal should be all requests serviced in <100ms. It's bothering me that I don't know where the slowness comes from, and that we have no good way of finding out. 

I'm thinking of a "query log" style log, which looks approximately like this:

```
{
  "url": "/bla",
  "time": 3185,
  "user": "asdf",
  "programs_loaded": 381,
  "db_time": 581,
  "bcrypt_time": 183,
  ...
}
```

Basically, a set of key-value pairs per requests which we can fill with all kinds of useful counts and timers and metadata that we can use to debug slow requests.

**Comment by fpereiro** [comment]
That's a brilliant suggestion! I will work on it right away.

**Comment by rix0rrr** [comment]
Actually started on it already if you don't mind :) holiday project

**Comment by fpereiro** [comment]
Sounds great!! I'll review the PR now :)

Should we close this issue or leave it open in case the profiling reveals we should change the schema?

**Comment by rix0rrr** [comment]
Nah. Let's close this.
